### PR TITLE
master|anthony/gemify-injectable

### PR DESCRIPTION
**Pull request by amrocco**
- [x] Copies the Injectable modules, structs, and classes from the Rubicon App to our new gem
- [x] Copies over the Injectable specs as well
- [x] A few changes to get things green
- [x] Updates the gemspec

--------------------
- [x] Specs pass

### CODE REVIEW

File: spec/injectable/dependencies_graph_spec.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-describe Injectable::DependenciesGraph, '#resolve' do
+RSpec.describe Injectable::DependenciesGraph, '#resolve' do
```
Comment (iovis9) [inline spec/injectable/dependencies_graph_spec.rb]: That is not necessary, if you need it then you have RSpec misconfigured

Comment (iovis9) [reply spec/injectable/dependencies_graph_spec.rb]: Also, that `'#resolve'` tag should not be there or in any of the others. I guess you need a new `describe` block

File: injectable.gemspec [hunk 2]
```diff
@@ -36,7 +36,10 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
 
-  spec.add_development_dependency "bundler", "~> 2.0"
-  spec.add_development_dependency "rake", "~> 10.0"
-  spec.add_development_dependency "rspec", "~> 3.0"
+  spec.add_runtime_dependency 'activesupport', '~> 5.2'
```
Comment (iovis9) [inline injectable.gemspec]: I'd like this dependency removed. We shouldn't need to use ActiveSupport for this gem.

File: injectable.gemspec [hunk 3]
```diff
@@ -36,7 +36,10 @@ Gem::Specification.new do |spec|
...
+
+  spec.add_development_dependency 'bundler', "~> 2.0"
+  spec.add_development_dependency 'rake', "~> 10.0"
+  spec.add_development_dependency 'rspec', "~> 3.0"
+  spec.add_development_dependency 'mocha', '~> 1.0'
```
Comment (iovis9) [inline injectable.gemspec]: We should rely in the native rspec mocks and move away from Mocha

File: injectable.gemspec [hunk 4]
```diff
@@ -36,7 +36,10 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
 
-  spec.add_development_dependency "bundler", "~> 2.0"
-  spec.add_development_dependency "rake", "~> 10.0"
-  spec.add_development_dependency "rspec", "~> 3.0"
+  spec.add_runtime_dependency 'activesupport', '~> 5.2'
+
+  spec.add_development_dependency 'bundler', "~> 2.0"
+  spec.add_development_dependency 'rake', "~> 10.0"
```
Comment (iovis9) [inline injectable.gemspec]: That's a very old rake version...

File: lib/injectable.rb [hunk 5]
```diff
@@ -5,6 +5,8 @@
 require 'injectable/dependency'
 require 'injectable/instance_methods'
 require 'injectable/missing_dependencies_exception'
+require 'active_support'
+require 'active_support/core_ext'
```
Comment (iovis9) [inline lib/injectable.rb]: Let's move away from these

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -8,6 +8,8 @@
   # Disable RSpec exposing methods globally on `Module` and `main`
   config.disable_monkey_patching!
 
+  config.mock_with :mocha
```
Comment (iovis9) [inline spec/spec_helper.rb]: And from these

File: injectable.gemspec [hunk 7]
```diff
@@ -4,42 +4,39 @@ $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
 require "injectable/version"
 
 Gem::Specification.new do |spec|
-  spec.name          = "injectable"
+  spec.name          = 'injectable'
   spec.version       = Injectable::VERSION
-  spec.authors       = ["amrocco"]
-  spec.email         = ["arocco89@gmail.com"]
+  spec.authors       = %w(Papipo iovis9 jantequera amrocco)
+  spec.email         = %w(papipo@gmail.com davidmarchan@gmail.com julio.antequera.galiano@gmail.com arocco89@gmail.com)
```
Comment (iovis9) [inline injectable.gemspec]: I'm not sure we should use our personal email addresses

File: injectable.gemspec (hunk 2 above)
Comment (amrocco) [reply injectable.gemspec]: That's the plan, but I want to get this set up as a gem first, then start removing dependencies in separate PRs.

File: injectable.gemspec [hunk 8]
```diff
@@ -4,42 +4,39 @@ $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
...
 
   # Prevent pushing this gem to RubyGems.org. To allow pushes either set the 'allowed_push_host'
   # to allow pushing to a single host or delete this section to allow pushing to any host.
   if spec.respond_to?(:metadata)
-    spec.metadata["allowed_push_host"] = "TODO: Set to 'http://mygemserver.com'"
-
-    spec.metadata["homepage_uri"] = spec.homepage
-    spec.metadata["source_code_uri"] = "TODO: Put your gem's public repo URL here."
-    spec.metadata["changelog_uri"] = "TODO: Put your gem's CHANGELOG.md URL here."
+    spec.metadata['homepage_uri'] = spec.homepage
```
Comment (iovis9) [inline injectable.gemspec]: Just remove this whole `if/else` statement

File: lib/injectable/version.rb [hunk 9]
```diff
@@ -1,3 +1,3 @@
 module Injectable
-  VERSION = "0.1.0"
+  VERSION = '1.0.0'
```
Comment (iovis9) [inline lib/injectable/version.rb]: When you change the version you have to update the lockfile as well, run `bundle install`

File: spec/injectable/dependencies_graph_spec.rb (hunk 1 above)
Comment (iovis9) [reply spec/injectable/dependencies_graph_spec.rb]: To configure `RSpec` to not need to do `RSpec.describe` you just remove this line: https://github.com/rubiconmd/injectable/blob/121944c53125cb98f35a68e4e984603b36b34a8d/spec/spec_helper.rb#L9

File: lib/injectable.rb [hunk 10]
```diff
@@ -1,6 +1,66 @@
...
+#     def team_must_accept_players!
+#       team.accepts_players? || raise TeamFullException
+#     end
+#   end
+#
+#   And use it like this:
+#
+#   AddPlayerToTeamRoster.call(player_id: player.id, team_id: team.id)
 module Injectable
   class Error < StandardError; end
```
Comment (iovis9) [inline lib/injectable.rb]: You don't really need that class

File: injectable.gemspec (hunk 2 above)
Comment (iovis9) [reply injectable.gemspec]: I'll have to live with that

File: Gemfile.lock [hunk 11]
```diff
@@ -0,0 +1,52 @@
+PATH
+  remote: .
+  specs:
+    injectable (0.1.0)
```
Comment (iovis9) [inline Gemfile.lock]: Once you run `bundle install` I'm guessing this one will change :D

File: spec/injectable/dependencies_graph_spec.rb (hunk 1 above)
Comment (amrocco) [reply spec/injectable/dependencies_graph_spec.rb]: Thanks!

File: lib/injectable/version.rb (hunk 9 above)
Comment (amrocco) [reply lib/injectable/version.rb]: Good to know, thanks!

**Comment by iovis9** [comment]
I see there are commits for Injectable in this PR: https://github.com/rubiconmd/rubicon/pull/2661/files
Are those changes here?

**Comment by iovis9** [comment]
I also found your personal email in the code of conduct document

**Comment by amrocco** [comment]
No, but that ticket is still waiting to go into qa.

**Comment by Papipo** [comment]
Even if it need to go through qa, the behaviour in that PR is the behaviour we probably want in Injectable. It has a smart dependency resolution mechanism which is quite handy.

**Pull request merged**